src/main/java/com/iota/iri/service/ledger/impl/LedgerServiceImpl.java

```diff
@@ -150,12 +150,12 @@ public boolean isBalanceDiffConsistent(Set<Hash> approvedHashes, Map<Hash, Long>
         Hash transactionPointer;
         while ((transactionPointer = nonAnalyzedTransactions.poll()) != null) {
             if (visitedTransactions.add(transactionPointer)) {
-
                 try {
                     final TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle,
                             transactionPointer);
 
-                    if (milestoneService.transactionBelongsToMilestone(transactionViewModel, milestoneIndex)) {
+                    if (milestoneService.transactionBelongsToMilestone(transactionViewModel,
+                            latestSolidMilestoneIndex + 1)) {
```

Comment:
I just have a little question:
In `MilestoneServiceImpl`:
```
 @Override
    public boolean transactionBelongsToMilestone(TransactionViewModel transaction, int milestoneIndex) {
        return transaction.snapshotIndex() == 0 || transaction.snapshotIndex() >= milestoneIndex;
    }
```

If transaction belongs to a milestone shouldn't it be  `transaction.snapshotIndex() <= milestoneIndex`?

Reply:
I think what you want to do after you fix `transactionBelongsToMilestone` is do:
`if (!milestoneService.transactionBelongsToMilestone(transactionViewModel,
                            latestSolidMilestoneIndex))` (note the **!**)

Reply:
Nope its not <= milestoneIndex because that would address ALL transactions that happened before the given index.

Lets say I want to retrieve all TX that belong to the milestone with index 10. Your equasion would return true for all tx with snapshotIndex 10, 9, 8, 7, 6, 5 ,4 ,3, 2 ,1 ... but we only want the ones that belong to exactly 10.

What we do here is pretty much the following.

Transactions that belong to milestone 10 are the ones that:

1. either have not been marked by a previous milestones yet ( == 0)
2. already carry the correct snapshot index ( == 10)
3. have mistakenly been marked by a future milestone ( > 10)

This can then be merged to == 0 || >= 10

This is utility method is only used while iterating the approovees from a milestone and it therefore needs to cover these 3 cases.

Reply:
Wait one second... In the old `LedgerValidator` we could have start traversing from an area that is non confirmed if I am not mistaken and calculate the balance (`getLatestDiff`). How do we do it now?

In the old version condition (3) is not accurate. In the new `LedgerService` it is since you start traversing from a `milestoneHash`.

Reply:
I see actually now. `milestoneHash` is a misleading name.

Reply:
Your logic works the problem is that definition (3) is not exactly correct in `LedgerServiceImpl`.
It is correct the way you use it in `MilestoneServiceImpl`...

We need to better define what this method does. In `LedgerServiceImpl` it is possible that you have a transaction that approved (indirectly) the latest milestone and in the meanwhile another thread assigned it a larger snapshot value.